hazelcast/src/main/java/com/hazelcast/query/impl/AbstractIndex.java

```diff
@@ -161,7 +161,8 @@ public void removeEntry(Data key, Object value, OperationSource operationSource)
 
         Set<Comparable> convertedValues = createHashSet(values.length);
         for (Comparable value : values) {
-            convertedValues.add(convert(value));
+            Comparable converted = convert(value);
+            convertedValues.add(canonicalizeQueryArgumentScalar(converted));
```

Comment:
Just to reiterate: ordered indexes will not canonicalize the values for storage as they don't support hash lookups. But this method is invoked from the `InPredicate` and here we are canonicalizing the query argument. Am I missing something?

Reply:
Yes, exactly. We need to perform that canonicalization here just to eliminate the duplicates from the passed set values, to avoid duplicate entries in the result, so any kind of an index may use `FastMultiResultSet` properly. At the same time, unordered/hash index stores provide `canonicalizeQueryArgumentScalar` implementations directly representing values canonically according to their storage format (`canonicalizeScalarForStorage`), so no additional canonicalization is required before querying the underlying store.

Basically, there are two cases: (1) ordered indexes, store values without canonicalization, canonicalization is required only for arguments of `InPredicate` to eliminate the duplicates (it's not necessary matches the stored values exactly, since in the end all queries are performed according to the comparison order);  (2) unordered indexes, store values in a canonical form, canonicalization is required for all arguments and it must match the canonicalization applied to stored values, otherwise equals/hashCode won't work properly.

Reply:
Got it, thanks!